packages/aws-cdk-lib/core/lib/adr/acknowledge-warnings.md

```diff
@@ -0,0 +1,181 @@
+```
+
+We will continue to use the node metadata as the storage mechanism and will add
+a new metadata type `aws:cdk:acknowledge` to store information on
+acknowledgements.
+
+At synthesis when we collect all of the annotation messages, we will filter out
+any messages that have an acknowledgement.
+
+Storing the acknowledgements in the metadata will also allow us to report on
+warnings that were acknowledged by the user (info will be stored in the
+assembly).
+
+## Alternatives
+
+### Alternative storage mechanism
+
+One alternative that was considered was to implement some alternative
+intermediary storage mechanism. This storage mechanism would allow storing all
+warnings and acknowledgements in a special construct that was created once per
+app. It would look something like this (pseudo code)
+
+```ts
+class AnnotationManager extends Construct {
+  private constructor(scope: Construct) {
+    attachCustomSynthesis(this, {
+      onSynthesize: () => {
+        this.warnings.forEach(warning => node.addMetadata(...))
+      }
+    }
+  }
+ 
+  public addWarning() {
+     if (!this.acks.has()) {
+       this.warnings.set();
+    }
+  }
+  public ack() {
+    if (this.warnings.has()) {
+      this.warnings.delete();
+    }
+    this.acks.add();
+  }
+}
+```
+
+The problem with this method is represented by the `attachCustomSynthesis` in
+the example. This same applies for if we used `addValidation` or `Aspects`.
+Annotations can be added _after_ that which means they would not be added or
+acknowledged.
+
+### Use context and remove metadata
+
+Another alternative that was considered was to use context to set
+acknowledgements. This would look something like this:
+
+```ts
+public acknowledgeWarning(id: string, message?: string) {
+  this.scope.node.setContext(id, message ?? true); // can't do this today
+  this.scope.node.removeMetadata(id); // this method does not exist
+}
+public addWarningV2(id: string, message: string) {
+  if (this.scope.node.tryGetContext(id) === undefined) {
+    this.addMessage(...);
+  }
+}
+```
+
+There are two issues with this alternative.
+
+1. It is currently not possible to `node.setContext` once children are added. We
+   would have to first remove that limitation. I think this could lead to a lot
+   of issues where users have to keep track of _when_ context is added.
+2. We would have to implement the ability to `removeMetadata` from a node. This
+   functionality doesn't make much sense outside of this context (can't find
+   where anybody has asked for it). It also may require updating the metadata
+   types since currently there is no unique identifier for a given metadata
+   entry (other than the message).
+
+## Consequences
+
+With the recommended solution the only major consequence is that it requires
+updating the `metadata-schema`, but we can do this in a non-breaking way
+(addition of new types). The alternatives may also require changes to the schema
```

Comment:
It changes the schema, and interpreting the schema becomes more complex. I would really really like to avoid having an add list and a scrap list in the same piece of data, that the client hen has to evaluate.

Adding this as a feature to the core seems preferable to me.

Reply:
@rix0rrr Then do you think the context + remove metadata option is the one to go with?